REGRESSION: ASSERTION FAILED: repaintRects() in WebCore::RenderLayer::recursiveUpdateLayerPositions.

mattwoodrow opened the pull request:
#### b37bf876c9b79cfe225c7858c23e58e8a2cd43e6
<pre>
REGRESSION: ASSERTION FAILED: repaintRects() in WebCore::RenderLayer::recursiveUpdateLayerPositions.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281041">https://bugs.webkit.org/show_bug.cgi?id=281041</a>
&lt;<a href="https://rdar.apple.com/137722495">rdar://137722495</a>&gt;

Reviewed by Simon Fraser.

The presence of repaintRects now depends on isVisibilityHiddenOrOpacityZero(),
not just hasVisibleContent.  Make sure we invalidate layer positions for opacity
changes, and change the logging to be the value that we care about.

Also switches the layer assertions to preferentially use debug asserts, so that
they are in use more often.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
(WebCore::outputLayerPositionTreeLegend):
(WebCore::outputLayerPositionTreeRecursive):

Canonical link: <a href="https://commits.webkit.org/285155@main">https://commits.webkit.org/285155@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96d11aef88bd6639f23507db8afcce238a6aa0ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71554 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50967 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24328 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75666 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22759 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73669 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58768 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22579 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56512 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14984 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74620 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46239 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61632 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36961 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42902 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19098 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21100 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64805 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19462 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77474 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15788 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18637 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64305 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15831 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61671 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64301 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15851 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12366 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6006 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46767 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1546 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47838 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49122 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47580 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5837,6 +5842,9 @@ void RenderLayer::styleChanged(StyleDifference diff, const RenderStyle* oldStyle
                 m_scrollableArea->computeHasCompositedScrollableOverflow(diff <= StyleDifference::RepaintLayer ? LayoutUpToDate::Yes : LayoutUpToDate::No);
         }
+
+        if (!oldStyle->opacity() != !renderer().style().opacity())
```
That looks odd, I'd rewrite it without the negations.

`if (oldStyle->isOpaque() != style().isOpaque())` and introduce a new `bool isOpaque() const` helper in RenderStyle, that, unlike `hasOpacity` exactly checks for opacity == 0.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5668,6 +5671,8 @@ void RenderLayer::updateSelfPaintingLayer()
         return;
 
     m_isSelfPaintingLayer = isSelfPaintingLayer;
+    setNeedsPositionUpdate();
```
Does that imply `isVisibilityHiddenOrOpacityZero` status changed? I did not entirely understand why you need this here.

nikolaszimmermann reviewed (commented):
Thanks @mattwoodrow for working on a fix! Still some questions, comments.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1103,12 +1103,15 @@ bool RenderLayer::ancestorLayerPositionStateChanged(OptionSet<UpdateLayerPositio
 #else
-#define LAYER_POSITIONS_ASSERT(assertion, ...) ASSERT(assertion, __VA_ARGS__)
-#define LAYER_POSITIONS_ASSERT_IMPLIES(condition, assertion) ASSERT_IMPLIES(condition, assertion)
+#define LAYER_POSITIONS_ASSERT(assertion, ...)
```
That change triggers empty body warnings.

```
/app/webkit/Source/WebCore/rendering/RenderLayer.cpp:1275:47: error: suggest braces around empty body in an ‘else’ statement [-Werror=empty-body]
 1275 |         LAYER_POSITIONS_ASSERT(dest == source);
```

Maybe: `do { } while (0)` ?

fujii replied:
`ASSERT` is using `((void)0)` for it.
https://github.com/WebKit/WebKit/blob/b4239a38c109a1f5980fc9c1cfe9a98cd3328741/Source/WTF/wtf/Assertions.h#L376

IMHO, CONJECTURE_ASSERT and ASSERT are different concepts.
It's weird that an ASSERT becomes a CONJECTURE_ASSERT if the build configuration defines ENABLE_CONJECTURE_ASSERT.

nikolaszimmermann replied:
Good idea to mimic what we already do. I just recently learned about the new conjecture mode. Others should comment on that.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5837,6 +5842,9 @@ void RenderLayer::styleChanged(StyleDifference diff, const RenderStyle* oldStyle
         }
+
+        if (oldStyle->hasZeroOpacity() != renderer().style().hasZeroOpacity())
+            setNeedsPositionUpdate();
```
The naming is odd; opacity has nothing to do with position.

mattwoodrow replied:
It's true, though I think the mismatch isn't here.

`setNeedsPositionUpdate` means 'we need to run `recursiveUpdateLayerPositions` for this layer.

Both are probably misnamed, since 'updating layer positions' also does a bunch of other non-position things, including computing repaint rects (which do depend on opacity).

mattwoodrow committed:
REGRESSION: ASSERTION FAILED: repaintRects() in WebCore::RenderLayer::recursiveUpdateLayerPositions.
https://bugs.webkit.org/show_bug.cgi?id=281041
<rdar://137722495>

Reviewed by Simon Fraser.

The presence of repaintRects now depends on isVisibilityHiddenOrOpacityZero(),
not just hasVisibleContent.  Make sure we invalidate layer positions for opacity
changes, and change the logging to be the value that we care about.

Also switches the layer assertions to preferentially use debug asserts, so that
they are in use more often.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
(WebCore::outputLayerPositionTreeLegend):
(WebCore::outputLayerPositionTreeRecursive):

Canonical link: https://commits.webkit.org/285155@main
